### PR TITLE
commit module 5 files

### PR DESCRIPTION
**Pull request by adrianadiaz02**

### CODE REVIEW

File: poetry.lock
Comment (pabaldonedo) [inline poetry.lock:1]: Why do we have changes in the .lock and not in the pyprojec.toml?

File: src/module_5/module5.md [hunk 1]
```diff
@@ -0,0 +1,1816 @@
...
+
+```python
+# Get top 500 stocks by market cap
+data_set = data_set.sort_values(by=["execution_date", "Market_cap__minus_365"], ascending=[True, False])
+top_500_stocks = data_set.groupby("execution_date").head(500)
+
+# Calculate the average surplus over the SP500 for the top 500 stocks
+benchmark_performance = top_500_stocks.groupby("execution_date").apply(
+    lambda x: (x["stock_change_div_365"] - x["sp500_change_365"]).mean()
+).reset_index(name="avg_surplus")
```
Comment (pabaldonedo) [inline src/module_5/module5.md:753]: fair baseline. For the record, normally, the base portfolio itself would be the S&P500 itself, where each stock is weighted by its market cap.

File: src/module_5/module5.md [hunk 2]
```diff
@@ -0,0 +1,1816 @@
...
+ggplot(combined_data) + geom_point(aes(x = "execution_date", y = "weighted-return", color='type')) + theme(axis_text_x = element_text(angle = 90, vjust = 0.5, hjust=1))
+```
+
+
+    
+![png](module5_files/module5_43_0.png)
+    
+
+
+We can observe that the returns are quite lower than the ones from the model given, and much more reasonable.
```
Comment (pabaldonedo) [inline src/module_5/module5.md:815]: I'd say they are more consisntet (Aka smooth) than just lower, which seems reasonable.

File: src/module_5/module5.md [hunk 3]
```diff
@@ -0,0 +1,1816 @@
...
+# Combine both
+combined_data = pd.concat([
+    benchmark_performance[['execution_date', 'weighted-return', 'type']],
+    filtered_results[['execution_date', 'weighted-return', 'type']]
+])
+```
+
+
+```python
+ggplot(combined_data) + geom_point(aes(x = "execution_date", y = "weighted-return", color='type')) + theme(axis_text_x = element_text(angle = 90, vjust = 0.5, hjust=1))
```
Comment (pabaldonedo) [inline src/module_5/module5.md:806]: Nice that you opted to try ggplot out

File: src/module_5/module5.md [hunk 4]
```diff
@@ -0,0 +1,1816 @@
...
+To diagnose if there is data leakage, we should usually:
+- Look at feature importance and understand if the most important features make sense
+- Retrain the model without suspicious features and see how much the performance decreases
+
+And to solve it we should:
+- Try to make a train/test split which resembles the production conditions
+- Make sure that all the information that the model has in offline mode is going to be available in online too
+
+
+Additionally, with the previous models, we perform feature importance:
```
Comment (pabaldonedo) [inline src/module_5/module5.md:848]: Split it in a different section, otherwise it seems convoluted

File: src/module_5/module5.md [hunk 5]
```diff
@@ -0,0 +1,1816 @@
...
+plt.show()
+```
+
+
+    
+![png](module5_files/module5_51_0.png)
+    
+
+
+We can observe that close_0 have abnormally high importance. It does not make sense that close_0 is the most important feature, as it represents what was the price at the moment of execution date. It would not make sense the idea of a model that basically defines whereas you should buy and action based on if the price is low or not.
```
Comment (pabaldonedo) [inline src/module_5/module5.md:893]: I have to say that this is pretty bad. There are too many plots (and they become small also) so the reader gets flooded with information and ends up not knowing where to look at. When presenting results try to spot what is relevant and focus on that so the others can follow easierly.

You can plot everything as dev, of course. But then, just state "we are showing the nost relevant plots/info below" and then repeat only what you really want to show and comment

File: src/module_5/module5.md [hunk 6]
```diff
@@ -0,0 +1,1816 @@
...
+
+It is worth mentioning that we should never trust the feature importance given by the trees, as it tends to give more importance to categorical features with a high number of categories. The more ideal approach for these cases is using permutation importance, which swaps columns and, if the performance degradates a lot, then the column is important.
+
+I will retrain the model without these two suspicious features and without the previously removed features and see how much the model performance decreases:
```
Comment (pabaldonedo) [inline src/module_5/module5.md:897]: which two other than close_0?

File: src/module_5/module5.md [hunk 7]
```diff
@@ -0,0 +1,1816 @@
...
+ggplot(train_results_final_tree) + geom_point(aes(x = "execution_date", y = "weighted-return")) + theme(axis_text_x = element_text(angle = 90, vjust = 0.5, hjust=1))
+```
+
+
+    
+![png](module5_files/module5_59_0.png)
+    
+
+
+With this change, we can see that the training results were similar: initial and last execution dates have 0 return and the middle ones have return distributed with values from 0.75 to 5.5. On the other hand, in the test set, the return is centered around 0, and it looks like the variance increases for future dates. Additionally, we can observe that we have now negative returns. It is important to see that during training, there are several dates were the return is very high (with values higher than 2), whereas in validation the return is not higher than 0.4. Even though, these two values are still higher than we would expect.
```
Comment (pabaldonedo) [inline src/module_5/module5.md:1084]: In the interest of time it is fine to leave it here. But with more time it can be good to understand what was causing the abnoral behviour since it can point to further problems

File: src/module_5/module5.md [hunk 8]
```diff
@@ -0,0 +1,1816 @@
...
+ggplot(train_results_final_tree) + geom_point(aes(x = "execution_date", y = "weighted-return")) + theme(axis_text_x = element_text(angle = 90, vjust = 0.5, hjust=1))
+```
+
+
+    
+![png](module5_files/module5_71_0.png)
+    
+
+
+We can observe that the returns are still not what we expect, so we will continue exploring.
```
Comment (pabaldonedo) [inline src/module_5/module5.md:1291]: This part of the code was meant to speed up the training process thus:

- I would expect execution time comparisons with current approach and old approch
- Performance differences between both approaches.

The goal would be to reduce first ponit while keeping the second untouch.

Since we are only planing to tackle the execution time, no real model improvement was expected beforehand.

File: src/module_5/module5.md [hunk 9]
```diff
@@ -0,0 +1,1816 @@
...
+
+#### 3 & 5. Try to understand if the algorithm is learning correctly & Try to see if there is overfitting
+
+With the high overfitting and the extremely high returns we can interpret the model is not learning correctly. To try to solve the model, we will change the hyperparameters. We will reduce the learning rate, to make smaller steps and probe more through the space. Particularlly, for trees, the learning rate controls how much we are going to weight each new tree compared to previous ones. Additionally, we will also reduce the number of estimators to try to avoid the overfitting, as each estimator (tree) added to the model increases its complexity, and the model starts to fit the noise in the training data rather than the underlying pattern.
```
Comment (pabaldonedo) [inline src/module_5/module5.md:1295]: There has not been a thorough exploration to state the first point. You have train and val plots that show big differences in performance but:

- Are these because the model is not learning? For example, learning rate too high that diverges, problems with the features, etc.
- Or because temporal shift in the distributions of train-test?


Regarding the number of trees, you can always fit more fits than optimum and then plot how error evolves with trees and prune

File: src/module_5/module5.md [hunk 10]
```diff
@@ -0,0 +1,1816 @@
...
+             "n_estimators":n_estimators,
+             "learning_rate": learning_rate,
+             "objective": objective,
+             "metric": metric}
+    
+    model = lgb.train(params = params,train_set = lgb_train,
+                      valid_sets = [lgb_test,lgb_train],
+                      feval = [top_wt_performance],
+                      callbacks = [lgb.record_evaluation(eval_result = eval_result)])
+    return model,eval_result,X_train,X_test
```
Comment (pabaldonedo) [inline src/module_5/module5.md:1368]: Although a notebook try to comply with PEP8, you will get used to it so it will become your default way of coding

File: src/module_5/module5.md [hunk 11]
```diff
@@ -0,0 +1,1816 @@
...
+ggplot(train_results_final_tree) + geom_point(aes(x = "execution_date", y = "weighted-return")) + theme(axis_text_x = element_text(angle = 90, vjust = 0.5, hjust=1))
+```
+
+
+    
+![png](module5_files/module5_82_0.png)
+    
+
+
+We can see that, for validations, the returns are now quite reasonable. Even though there are two peaks, the other execution dates have returns not higher than 0.2 and we also have some negative returns. During training, the returns are still quite high but not has much as at the begining.
```
Comment (pabaldonedo) [inline src/module_5/module5.md:1454]: Why? what is reasonable? Always define what is "normal", "abnormal", "reasonable" or "suspicious". Never assume this since different people may have different ideas in their head and you may end up discussing with someone with who you share vocabulary but not meaning

File: src/module_5/module5.md [hunk 12]
```diff
@@ -0,0 +1,1816 @@
...
+    No model returned for execution date: 2021-03-27T00:00:00.000000000
+
+
+
+```python
+test_results = parse_results_into_df("valid_0")
+train_results = parse_results_into_df("training")
+
+test_results_final_tree = test_results.sort_values(["execution_date","n_trees"]).drop_duplicates("execution_date",keep = "last")
+train_results_final_tree = train_results.sort_values(["execution_date","n_trees"]).drop_duplicates("execution_date",keep = "last")
```
Comment (pabaldonedo) [inline src/module_5/module5.md:1429]: You are sticking with 20 trees, but how do you know that is a good choice? Maybe 10 is better, or 25... plot error evolution with trees to find out

File: src/module_5/module5.md [hunk 13]
```diff
@@ -0,0 +1,1816 @@
...
+    
+![png](module5_files/module5_82_0.png)
+    
+
+
+We can see that, for validations, the returns are now quite reasonable. Even though there are two peaks, the other execution dates have returns not higher than 0.2 and we also have some negative returns. During training, the returns are still quite high but not has much as at the begining.
+
+#### 4. We are using a very high level metric to evaluate the algorithm so you maybe need to use some more low level ones
+
+My first idea was to use the typical metrics precision, recall, roc_auc, and mae. Buy actually, this is not what was expected. The thing here is that, even tough it is important to have the final business metric (weighted return) in mind, the actual metric the model is trying to minimize is the binary log-loss. Therefore, in order to understand what the model is doing, we should work with this metric, which is closer to the model.
```
Comment (pabaldonedo) [inline src/module_5/module5.md:1458]: Small comment: there is not anything "expected". The point is always to give you tools and what is "expected" is to become independent and professional in problem solving using data approaches

File: src/module_5/module5.md [hunk 14]
```diff
@@ -0,0 +1,1816 @@
...
+plt.xlabel('Tree Number')
+plt.ylabel('Percentage Difference of Binary Log-Loss')
+plt.title('Percentage Differences of Validation Binary Log-Loss Across Trees')
+plt.grid()
+plt.show()
+```
+
+
+    
+![png](module5_files/module5_90_0.png)
```
Comment (pabaldonedo) [inline src/module_5/module5.md:1578]: The blue fill makes it hard to visualise when we are improving (Below 0) and when we are not

File: src/module_5/module5.md [hunk 15]
```diff
@@ -0,0 +1,1816 @@
...
+    
+
+
+The box plot analysis reveals that while the model improves with the addition of trees initially, there is a point (more or less 7 trees) after which additional trees introduce more variability and potential overfitting, as even though the mean continues to decrease, it has the drawback of increasing variability.
```
Comment (pabaldonedo) [inline src/module_5/module5.md:1582]: It is hard for me to understand why you picked 7 and not any other number using variance and mean. Try to be systematic and with well and easy to justify decision making

**Review by pabaldonedo — APPROVED** [review]
I can see a lot of good ideas and approaches scatter along the notebook. However, I feel it requires a little bit more of order and better justification of the decisions made. For example, after finishing reading the notebook I would not be able to clearly explain what have been the gains of the new hyperparmeters vs reducing number of trees vs reducing number of stocks. It is there accross the notebook but not clearly and properly structured. In addition, some decisions are not explict enough on the rationale that made them happen, for example, the number of trees.

About the improving execution time section, you focused at the end in predictive performance instead of the execuition time, whicha was the main goal of the section.

This is perfectly normal when learning. I will suggest you to read the notebook on one go and try to think how would you restructure / add explanation to create a "story" that is easy to grasp that conducts the reader through all your work and ends up with him able to answer:

 - What leakages were discovered?
 - How faster is to reduce the dataset and how much it impact in performance?
 - How much changing the hyperparmeters affects to the error?
 - How many trees were selected and why and what was it boost?
 - How all these intermediate models and the final one compares to baseline?

Keep up the good work!
